src/Control/FunFlow/External/Coordinator.hs

```diff
@@ -60,6 +67,11 @@ class Coordinator c where
   popTask :: MonadIO m => Hook c -> Executor
           -> m (Maybe TaskDescription)
 
+  -- | Await task completion. If the task is not running or an invalid
+  --   content hash is submitted, this will return Nothing immediately.
+  --   If the task has failed, this will also return 'Nothing'.
+  awaitTask :: MonadIO m => Hook c -> ContentHash -> m (Maybe ExecutionInfo)
```

Comment:
I'm confused by this method's name and its documentation.

> If the task is not running [...] this will return Nothing immediately.

I can see two cases how a task can be not running. Either it hasn't started, yet, or it has finished.
If it's not running, yet, I would expect that a function called `awaitTask` would wait until it starts running and then wait further until it finishes.
If it's already finished, then I would expect `awaitTask` to just hand back the result.

> If [...] an invalid content hash is submitted, this will return Nothing immediately.
> If the task has failed, this will also return 'Nothing'.

It seems to me that it would be cleaner to distinguish these two error conditions.

If the content hash is invalid, then the workflow interpreter messed up, right?
That seems like a very different error condition then task failure.
If the task just failed, that's an expected possible outcome and the caller will want to be able to handle it somehow. Either retry, or display an error.

Maybe instead return `m (Either FailureInfo ExecutionInfo)`, where `FailureInfo` holds `ExecutionInfo` and the exit code, and throw if an invalid content hash is submitted?

Reply:
All good points, and I knew this function smelled when I wrote it. I'll update it tomorrow!